crates/futures/src/polyfill.rs

```diff
@@ -0,0 +1,208 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ * Author: Lars T Hansen, lhansen@mozilla.com
+ */
+
+/* Polyfill for Atomics.waitAsync() for web browsers.
+ *
+ * Any kind of agent that is able to create a new Worker can use this polyfill.
+ *
+ * Load this file in all agents that will use Atomics.waitAsync.
+ *
+ * Agents that don't call Atomics.waitAsync need do nothing special.
+ *
+ * Any kind of agent can wake another agent that is sleeping in
+ * Atomics.waitAsync by just calling Atomics.wake for the location being slept
+ * on, as normal.
+ *
+ * The implementation is not completely faithful to the proposed semantics: in
+ * the case where an agent first asyncWaits and then waits on the same location:
+ * when it is woken, the two waits will be woken in order, while in the real
+ * semantics, the sync wait will be woken first.
+ *
+ * In this polyfill Atomics.waitAsync is not very fast.
+ */
+
+/* Implementation:
+ *
+ * For every wait we fork off a Worker to perform the wait.  Workers are reused
+ * when possible.  The worker communicates with its parent using postMessage.
+ */
+
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use js_sys::{
+    encode_uri_component, Array, Atomics, Error, Function, Int32Array, JsString, Promise, Reflect,
+    SharedArrayBuffer, WebAssembly,
+};
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::JsCast;
+use web_sys::{MessageEvent, Worker};
+
+macro_rules! console_log {
+    ($($t:tt)*) => (log(&format_args!($($t)*).to_string()))
+}
+
+#[wasm_bindgen]
+extern "C" {
+    #[wasm_bindgen(js_namespace = console)]
+    fn log(s: &str);
+}
+
+const DEFAULT_TIMEOUT: f64 = 10.0;
+
+const HELPER_CODE: &'static str = "
+onmessage = function (ev) {
+    try {
+        switch (ev.data[0]) {
+            case 'wait': {
+                let [_, ia, index, value, timeout] = ev.data;
+                console.log('wait event inside a worker');
+                let result = Atomics.wait(ia, index, value, timeout);
+                postMessage(['ok', result]);
+                break;
+            }
+            default: {
+                throw new Error('Wrong message sent to wait helper: ' + ev.data.join(','));
+            }
+        }
+    } catch (e) {
+        console.log('Exception in wait helper');
+        postMessage(['error', 'Exception']);
+    }
+}
+";
+
+thread_local! {
+    static HELPERS: RefCell<Vec<Rc<Worker>>> = RefCell::new(vec![]);
+}
+
+fn alloc_helper() -> Rc<Worker> {
+    HELPERS.with(|helpers| {
+        if let Some(helper) = helpers.borrow_mut().pop() {
+            return helper;
+        }
+
+        let mut initialization_string = "data:application/javascript,".to_owned();
+        let encoded: String = encode_uri_component(HELPER_CODE).into();
+        initialization_string.push_str(&encoded);
+
+        return Rc::new(Worker::new(&initialization_string).expect("Should create a Worker"));
```

Comment:
Sorry this may have gotten lost in the previous review comments, but I think that the `Rc` here shouldn't be necessary because a `Worker` can be natively cloned?